File: julia/Day7.jl
module Day7

export Submarine, on_input!, output

mutable struct Submarine
    crab_positions::Vector{Int32}

    Submarine() = new([])
end

function on_input!(s::Submarine, a::AbstractString)
    s.crab_positions = sort(parse.(Int32, split(a, ",")))
end

function min_assumed_fuel(s::Submarine)
    optimal_position = s.crab_positions[1 + length(s.crab_positions) ÷ 2]

    sum(map(pos -> abs(optimal_position - pos), s.crab_positions))
end

function min_real_fuel(s::Submarine)
    optimal_position = 1 + sum(s.crab_positions) ÷ length(s.crab_positions)

    minimum(map(pos -> real_fuel_for_pos(s, pos), (optimal_position - 1):(optimal_position + 1)))
end

function real_fuel_between_pos(pos1, pos2)::Int32
    diff = abs(pos1 - pos2)

    (diff * (diff + 1)) / 2
end

function real_fuel_for_pos(s::Submarine, pos)
    sum(map(x -> real_fuel_between_pos(pos, x), s.crab_positions))
end

function output(s::Submarine)
    println("Part 1: $(min_assumed_fuel(s))")
    println("Part 2: $(min_real_fuel(s))")
end

end

File: rust/src/day7.rs
use std::collections::HashMap;

pub struct Submarine {
    crab_positions: HashMap<i32, i32>,
    min_crab_position: i32,
    max_crab_position: i32,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            crab_positions: HashMap::new(),
            min_crab_position: i32::MAX,
            max_crab_position: 0,
        }
    }

    pub fn on_input(&mut self, result: &str) {
        result.split(",")
            .map(|x| x.parse().unwrap())
            .for_each(|x: i32| {
                if x < self.min_crab_position {
                    self.min_crab_position = x;
                }

                if x > self.max_crab_position {
                    self.max_crab_position = x;
                }

                *self.crab_positions.entry(x).or_default() += 1;
            });
    }

    fn min_assumed_fuel_for_positioning(&self) -> i32 {
        (self.min_crab_position..=self.max_crab_position)
            .map(|currpos| {
                self.crab_positions.iter().fold(0, |fuel, (pos, num_crabs)| fuel + (currpos - pos).abs() * num_crabs)
            })
            .min()
            .unwrap()
    }

    fn min_real_fuel_for_positioning(&self) -> i32 {
        (self.min_crab_position..=self.max_crab_position)
            .map(|currpos| {
                self.crab_positions.iter().fold(0, |fuel, (pos, num_crabs)| fuel + fuel_for_pos_difference(currpos - pos) * num_crabs)
            })
            .min()
            .unwrap()
    }

    pub fn output(&mut self) {
        let min_assumed_fuel = self.min_assumed_fuel_for_positioning();
        let min_real_fuel = self.min_real_fuel_for_positioning();

        println!("Part 1: {}", min_assumed_fuel);
        println!("Part 2: {}", min_real_fuel);
    }
}

fn fuel_for_pos_difference(mut pos_difference: i32) -> i32 {
    pos_difference = pos_difference.abs();

    (pos_difference * (pos_difference + 1)) / 2
}